File: CPP/algorithms/mathematical/euclidean_algorithm.cpp
/*
 * Algorithm: Euclidean Algorithm for GCD
 * Description: The Euclidean algorithm is an efficient method to compute the greatest common divisor (GCD) of two integers.
 *              It's based on the principle that if a and b are two positive integers, then gcd(a,b) = gcd(b, a mod b).
 * Time Complexity: O(log(min(a,b)))
 * Space Complexity: O(log(min(a,b))) for recursive approach, O(1) for iterative approach
 * Author: Abhiram
 */

#include <iostream>

// Recursive implementation of the Euclidean Algorithm
int gcdRecursive(int a, int b) {
    // Base case
    if (b == 0)
        return a;
    
    // Recursive case: gcd(a,b) = gcd(b, a mod b)
    return gcdRecursive(b, a % b);
}

// Iterative implementation of the Euclidean Algorithm
int gcdIterative(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

// Extended Euclidean Algorithm to find coefficients x and y such that ax + by = gcd(a,b)
void extendedGCD(int a, int b, int &x, int &y, int &gcd) {
    // Base case
    if (b == 0) {
        x = 1;
        y = 0;
        gcd = a;
        return;
    }
    
    // Recursive call
    int x1, y1;
    extendedGCD(b, a % b, x1, y1, gcd);
    
    // Update x and y using results of recursive call
    x = y1;
    y = x1 - (a / b) * y1;
}

int main() {
    int a, b;
    std::cout << "Enter two positive integers: ";
    std::cin >> a >> b;
    
    // Ensure positive integers
    a = abs(a);
    b = abs(b);
    
    // Calculate GCD using both methods
    std::cout << "GCD using recursive Euclidean algorithm: " << gcdRecursive(a, b) << std::endl;
    std::cout << "GCD using iterative Euclidean algorithm: " << gcdIterative(a, b) << std::endl;
    
    // Calculate coefficients using Extended Euclidean algorithm
    int x, y, gcd;
    extendedGCD(a, b, x, y, gcd);
    std::cout << "Extended Euclidean Algorithm results:" << std::endl;
    std::cout << "GCD: " << gcd << std::endl;
    std::cout << "Coefficients x and y: " << x << " and " << y << std::endl;
    std::cout << "Verification: " << a << " * " << x << " + " << b << " * " << y << " = " << (a*x + b*y) << std::endl;
    
    return 0;
}

File: Python/algorithms/mathematical/euclidean_algorithm.py
# 📌 Euclidean Algorithm for Greatest Common Divisor (GCD)

"""
Algorithm: Euclidean Algorithm
Description: Computes the greatest common divisor (GCD) of two non-negative integers.
Time Complexity: O(log(min(a, b))) - The number of steps is proportional to the number of digits in the smaller number.
Space Complexity: O(log(min(a, b))) for recursive calls, O(1) for iterative.
Author: Gemini
"""

def gcd_recursive(a: int, b: int) -> int:
    """
    Computes the greatest common divisor (GCD) of two non-negative integers using the recursive Euclidean algorithm.

    Args:
        a (int): The first non-negative integer.
        b (int): The second non-negative integer.

    Returns:
        int: The greatest common divisor of a and b.

    Raises:
        ValueError: If either a or b is negative.
    """
    if a < 0 or b < 0:
        raise ValueError("Both numbers must be non-negative.")
    if b == 0:
        return a
    return gcd_recursive(b, a % b)

def gcd_iterative(a: int, b: int) -> int:
    """
    Computes the greatest common divisor (GCD) of two non-negative integers using the iterative Euclidean algorithm.

    Args:
        a (int): The first non-negative integer.
        b (int): The second non-negative integer.

    Returns:
        int: The greatest common divisor of a and b.

    Raises:
        ValueError: If either a or b is negative.
    """
    if a < 0 or b < 0:
        raise ValueError("Both numbers must be non-negative.")
    while b:
        a, b = b, a % b
    return a

def main():
    """Test the Euclidean algorithm with example cases."""
    test_cases = [
        (48, 18, 6),
        (101, 103, 1),
        (0, 5, 5),
        (7, 0, 7),
        (10, 10, 10),
        (1, 1, 1),
        (12, 15, 3),
        (270, 192, 6)
    ]

    print("--- Recursive GCD ---")
    for a, b, expected in test_cases:
        try:
            result = gcd_recursive(a, b)
            print(f"GCD_recursive({a}, {b}) = {result} (Expected: {expected}) {'✅' if result == expected else '❌'}")
        except ValueError as e:
            print(f"GCD_recursive({a}, {b}) raised error: {e} (Expected: {expected}) ❌")

    print("\n--- Iterative GCD ---")
    for a, b, expected in test_cases:
        try:
            result = gcd_iterative(a, b)
            print(f"GCD_iterative({a}, {b}) = {result} (Expected: {expected}) {'✅' if result == expected else '❌'}")
        except ValueError as e:
            print(f"GCD_iterative({a}, {b}) raised error: {e} (Expected: {expected}) ❌")

    # Test edge cases with negative numbers
    print("\n--- Negative Number Test Cases ---")
    negative_test_cases = [
        (-5, 10),
        (10, -5),
        (-5, -10)
    ]
    for a, b in negative_test_cases:
        try:
            gcd_recursive(a, b)
            print(f"GCD_recursive({a}, {b}) - Expected ValueError, but no error. ❌")
        except ValueError as e:
            print(f"GCD_recursive({a}, {b}) - Caught expected error: {e} ✅")
        try:
            gcd_iterative(a, b)
            print(f"GCD_iterative({a}, {b}) - Expected ValueError, but no error. ❌")
        except ValueError as e:
            print(f"GCD_iterative({a}, {b}) - Caught expected error: {e} ✅")


if __name__ == "__main__":
    main()